File: frontend/src/api/reports.ts
import { apiRequest, resolveApiUrl } from "@/api/client";
import { getAdminToken } from "@/lib/adminSession";
import type {
  AbsentCandidateRow,
  QuestionAccuracyRow,
  ScoreReportRow,
  WrongQuestionRow,
} from "@/types/report";

function withExamFilter(path: string, examId?: string | null) {
  if (!examId) {
    return path;
  }
  const params = new URLSearchParams({ exam_id: examId });
  return `${path}?${params.toString()}`;
}

export function getScoreReport(examId?: string | null) {
  return apiRequest<ScoreReportRow[]>(withExamFilter("/api/admin/reports/scores", examId));
}

export function getQuestionAccuracy(examId?: string | null) {
  return apiRequest<QuestionAccuracyRow[]>(
    withExamFilter("/api/admin/reports/question-accuracy", examId),
  );
}

export function getWrongQuestions(examId?: string | null) {
  return apiRequest<WrongQuestionRow[]>(
    withExamFilter("/api/admin/reports/wrong-questions", examId),
  );
}

export type AttendanceStatus = "not_started" | "in_progress" | "submitted";

export function getAbsentCandidates(
  status: AttendanceStatus = "not_started",
  examId?: string | null,
) {
  const params = new URLSearchParams({ status });
  if (examId) {
    params.set("exam_id", examId);
  }
  return apiRequest<AbsentCandidateRow[]>(`/api/admin/reports/absent-candidates?${params}`);
}

export async function downloadReportExport(examId?: string | null): Promise<void> {
  const response = await fetch(resolveApiUrl(withExamFilter("/api/admin/reports/export", examId)), {
    headers: { "X-Admin-Token": getAdminToken() ?? "" },
  });
  if (!response.ok) throw new Error("报表导出失败");
  const blob = await response.blob();
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = "考试报表.xlsx";
  a.click();
  URL.revokeObjectURL(url);
}

File: backend/app/api/reports.py
from urllib.parse import quote

from fastapi import APIRouter, Depends
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.schemas.common import ApiResponse
from app.schemas.report import (
    AbsentCandidateRow,
    QuestionAccuracyRow,
    ScoreReportRow,
    WrongQuestionRow,
)
from app.services import report_service

router = APIRouter(prefix="/admin/reports", tags=["admin-reports"])


@router.get("/scores", response_model=ApiResponse[list[ScoreReportRow]])
def get_score_report(
    db: Session = Depends(get_db),
) -> ApiResponse[list[ScoreReportRow]]:
    return ApiResponse(data=report_service.get_score_report(db))


@router.get("/question-accuracy", response_model=ApiResponse[list[QuestionAccuracyRow]])
def get_question_accuracy(
    db: Session = Depends(get_db),
) -> ApiResponse[list[QuestionAccuracyRow]]:
    return ApiResponse(data=report_service.get_question_accuracy(db))


@router.get("/wrong-questions", response_model=ApiResponse[list[WrongQuestionRow]])
def get_wrong_questions(
    db: Session = Depends(get_db),
) -> ApiResponse[list[WrongQuestionRow]]:
    return ApiResponse(data=report_service.get_wrong_questions(db))


@router.get("/absent-candidates", response_model=ApiResponse[list[AbsentCandidateRow]])
def get_absent_candidates(
    exam_id: int | None = None,
    status: str = "not_started",
    db: Session = Depends(get_db),
) -> ApiResponse[list[AbsentCandidateRow]]:
    return ApiResponse(
        data=report_service.get_absent_candidates(db, exam_id=exam_id, status=status)
    )


@router.get("/export")
def export_report(db: Session = Depends(get_db)) -> StreamingResponse:
    stream = report_service.generate_report_workbook(db)
    return StreamingResponse(
        stream,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={
            "Content-Disposition": f"attachment; filename*=UTF-8''{quote('考试报表.xlsx')}"
        },
    )
